mchf-eclipse/drivers/audio/audio_convolution.h

```diff
@@ -0,0 +1,72 @@
+/*  -*-  mode: c; tab-width: 4; indent-tabs-mode: t; c-basic-offset: 4; coding: utf-8  -*-  */
+/************************************************************************************
+ **                                                                                **
+ **                                        UHSDR                                   **
+ **               a powerful firmware for STM32 based SDR transceivers             **
+ **                                                                                **
+ **--------------------------------------------------------------------------------**
+ **                                                                                **
+ **  Description:   Code for fast convolution filtering                            **
+ **  Licence:		GNU GPLv3                                                      **
+ ************************************************************************************/
+
+/* Define to prevent recursive inclusion -------------------------------------*/
+#ifndef __AUDIO_CONVOLUTION_H
+#define __AUDIO_CONVOLUTION_H
+
+#include "arm_math.h"
+//#include "softdds.h"
+#include "uhsdr_hw_i2s.h"
+#include "uhsdr_board.h"
+#include "codec.h"
+#include "audio_driver.h"
+#include "radio_management.h"
+#include "usbd_audio_if.h"
+
+extern arm_fir_decimate_instance_f32   DECIMATE_RX_I;
+extern arm_fir_decimate_instance_f32   DECIMATE_RX_Q;
+//extern AudioDriverBuffer adb;
+
+#ifdef USE_CONVOLUTION
+#define FFT_CONVOLUTION_SIZE 256
+#define CONVOLUTION_MAX_NO_OF_BLOCKS 8
+#define CONVOLUTION_MAX_NO_OF_COEFFS 2048
+#endif
+
+#ifdef USE_CONVOLUTION
+
+typedef struct
+{
+    // for convolution filtering
+    float32_t				i_buffer_convolution[FFT_CONVOLUTION_SIZE / 2];
+    float32_t				q_buffer_convolution[FFT_CONVOLUTION_SIZE / 2];
+
+    float32_t				fmask[CONVOLUTION_MAX_NO_OF_BLOCKS][FFT_CONVOLUTION_SIZE * 2];
+    float32_t				fftin[FFT_CONVOLUTION_SIZE * 2];
+    float32_t				fftout[CONVOLUTION_MAX_NO_OF_BLOCKS][FFT_CONVOLUTION_SIZE * 2];
+    float32_t				ifftout[FFT_CONVOLUTION_SIZE * 2];
+    float32_t				accum[FFT_CONVOLUTION_SIZE * 2];
+    float32_t               a_buffer[2][FFT_CONVOLUTION_SIZE / 2]; // for convolution, we need an output buffer of 128 samples
+} ConvolutionBuffers;
+
+//extern ConvolutionBuffers cob;
+static ConvolutionBuffers cob;
+
+typedef struct
+{
+    // for convolution filtering
+    int						nc; // no. of coefficients
+    int 					size; // no. of input samples
+    int						nfor; // no. of blocks in the convolution
+    int						buffidx; // buffer pointer
+    float32_t				impulse[CONVOLUTION_MAX_NO_OF_COEFFS * 2]; // impulse response has real and imaginary components
+    float32_t				maskgen[FFT_CONVOLUTION_SIZE * 2];
+} ConvolutionBuffersshared;
+
+extern ConvolutionBuffersshared cbs;
+
```

Comment:
You still need to put 
 `ConvolutionBuffersshared cbs;` in one of the .c files, so that the compiler knows, that it should really create the buffer in the context of this file.

`extern` says basically:  A variable with the given name and type will be defined somewhere by someone (that is called "declaring" the existence of a variable with a type). It does not take care that the variable is in fact created (i.e. "defined", meaning it becomes part of a compiled .o file). Only the linker will find this out (and giving you the error message for it).

Put above statement  `ConvolutionBuffersshared cbs;` in "audio_convolution.c" and leave `extern ConvolutionBuffersshared cbs;` in "audio_convolution.h" This is the pattern how you tell the "world" about a global variable (and its type) you defined in a .c file.

Reply:
thanks a lot, Danilo! that clears up the sky!
Fixed it quickly, but will have another close look at those things tomorrow with a (hopefully) clear brain.
73 Frank